File: solutions/python/perfect-numbers/1/perfect_numbers.py
from math import isqrt

def classify(number):
    """ A perfect number equals the sum of its positive divisors.

    :param number: int a positive integer
    :return: str the classification of the input integer
    """
    if number <= 0:
        raise ValueError("Classification is only possible for positive integers.")
    match sum((
        0 if divisor == number else
            divisor if dividend == number or dividend == divisor else
            divisor + dividend
        for divisor, dividend in (
            (divisor, number / divisor)
            for divisor in range(1, isqrt(number) + 1)
            if number % divisor == 0
        )
    )):
        case aliquot if aliquot > number:
            print(aliquot)
            return "abundant"
        case aliquot if aliquot == number:
            print(aliquot)
            return "perfect"
        case aliquot if aliquot < number:
            print(aliquot)
            return "deficient"

File: solutions/haskell/perfect-numbers/1/src/PerfectNumbers.hs
module PerfectNumbers (classify, Classification(..)) where

data Classification = Deficient | Perfect | Abundant deriving (Eq, Show)

classify :: Int -> Maybe Classification
classify i
  | i <= 0 = Nothing
  | otherwise = Just (if aliquot == i then Perfect else if aliquot > i then Abundant else Deficient)
    where aliquot = sum [x | x <- [1..(i - 1)], i `mod` x == 0]